### ISSUE TITLE
[request] Environmental variable containing the number of cores, threads, or similar

### ISSUE THREAD

**Issue by ariscop** [travis-ci/travis-ci#4696]
Would be handy for parallel make.
VM's are supposed to have two cores but i'd rather derive that from the environment than hardcode

**Comment by BanzaiMan** [comment]
In the virtualized environments, I do not think these values can be reliably discovered from within the VM.

**Comment by BanzaiMan** [comment]
The build script is compiled on a server that does not have any idea of the runtime environment. These values, then, must be discovered by the build script at run time. If there are commands to do that from VM, however, they should also be available to your build script at runtime.

**Comment by thedrow** [comment]
@BanzaiMan The VMs (at least the OpenVZ ones) report that there are 8 CPUs on the machine while there are only two. I have no idea why but that was the case last time I checked. Try running ```py.test -n auto``` with pytest-xdist installed and see for yourself.

**Comment by BanzaiMan** [comment]
@thedrow Right. I have not found a way to discover these values reliably in virtual environments. If there are, I would be happy to work on exposing the information to the build script.

**Comment by thedrow** [comment]
What happens if you run cat /proc/cpuinfo inside an OpenVZ container? I think you're not limiting the number of CPUs but configuring the CPU time shared between OpenVZ containers which is why all containers report that they have 8 CPUs. I'm just guessing here though.

**Comment by BanzaiMan** [comment]
That does not work. As I stated previously, the build script is compiled on a machine that does not necessarily run the build script itself. The value needs to be discoverable from the VM.

**Comment by thedrow** [comment]
@BanzaiMan Read again what I said. What happens when you run ```cat /proc/cpuinfo``` from inside the container/VM that's running the build?

**Comment by joshk** [comment]
From my understanding, there is a limitation with how LXC and OpenVZ works where asking the system for this info will give us the host system details, instead of the per container details.

We will have a think about this request and see how it fits into our 1-3 month plan which I am working on.

Thanks.

**Comment by BanzaiMan** [comment]
@thedrow Yes, I tried it, and that's not accurate.

**Comment by thedrow** [comment]
@joshk With any cgroups based container (LXC, docker etc.) you can get the number of effective CPUs with:
```
cat /sys/fs/cgroup/cpuset/system.slice/<container name>/cpuset.effective_cpus
```
That's at least the ubuntu file system format with Docker.

**Comment by ariscop** [comment]
I was imagining something done when the vm is provisioned, setting NUMCORES=2 in /etc/environment or similar

**Comment by thedrow** [comment]
@ariscop I don't think it makes it correct 100% of the times. You can get 1.5 dedicated cores.
Also this doesn't help with components that use Python/Ruby or any other programming language other than bash/shell.

**Comment by ariscop** [comment]
@thedrow Not quite sure what you mean, Python/Ruby can read from the environment?

My use case is very limited: Want to run make in parallel but don't know how many threads are available to my vm/container. If cpu time is capped to 1.5 threads worth, then I'd want to run 2 jobs.

**Comment by thedrow** [comment]
When you run ```python -c 'import multiprocessing; print(multiprocessing.cpu_count())'``` you get the wrong value because as @joshk mentioned this is a limitation of LXC/OpenVZ.
Both Python and Ruby or any other language that runs on Linux for that matter parses the CPUs count from /proc/cpuinfo which is left unmodified under when you are inside a container.
Adding an environment variable that always has the value of 2 has no value. Most of the typical parallel runs try to automatically infer how many processes to use based on the CPUs count so they will attempt to read from /proc/cpuinfo.
If you want to specify it explicitly then just run the tests with 2 threads/processes. It's not going to chance tomorrow so no worries there.
If you still want to get the number of effective CPUs assigned for your process or session you'll have to be more familiar with the /sys and /proc file systems.

**Comment by ariscop** [comment]
Oh, you mean to make existing detection work, that would need a different solution

**Comment by jayvdb** [comment]
All Travis environments types currently provide two cores, according to https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments .

In Python, one way to obtain the number of cores available for a Travis job, other than hard-coding '2', is to use [`psutil`](https://github.com/giampaolo/psutil)'s [`Process.cpu_affinity`](https://pythonhosted.org/psutil/#psutil.Process.cpu_affinity).  On travis the following gives the correct answer

```
>>> len(psutil.Process().cpu_affinity())
2
```

`multiprocessing.cpu_count()`, `os.cpu_count()`, etc all give 32, and even `psutil.cpu_count(physical=True)` returns 16.

**Comment by galeksandrp** [comment]
I've seen issue that questions why are docker builds does not have root. Possibility to do taskset 0xFFFFFFFF make -j32 looks like answer.

**Comment by ilanbiala** [comment]
Any update on this?

**Comment by ibtaylor** [comment]
Not sure if this is relevant, but I often will make use of the `nproc` tool from `coreutils` and do something hacky like this:

```bash
DEFAULT_NUM_MAKE_JOBS=2
NUM_MAKE_JOBS=$(($(nproc 2> /dev/null || echo ${DEFAULT_NUM_MAKE_JOBS})+1))
make -j${NUM_MAKE_JOBS}
```
